File: Backend/udp-server/src/main/java/net/chrisparton/sparkled/udpserver/UdpServerImpl.java
package net.chrisparton.sparkled.udpserver;

import com.google.common.util.concurrent.ThreadFactoryBuilder;

import javax.inject.Inject;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.SocketException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

public class UdpServerImpl implements UdpServer {

    private static final Logger logger = Logger.getLogger(UdpServerImpl.class.getName());
    private static final int THREAD_COUNT = 4;
    private static final int RECEIVE_BUFFER_SIZE = 16;
    private final RequestHandler requestHandler;
    private final ExecutorService executor;
    private boolean started;

    @Inject
    public UdpServerImpl(RequestHandler requestHandler) {
        this.requestHandler = requestHandler;
        this.executor = Executors.newFixedThreadPool(THREAD_COUNT,
                new ThreadFactoryBuilder().setNameFormat("udp-server-%d").build()
        );
    }

    @Override
    public void start(int port) throws SocketException {
        if (started) {
            logger.warning("Attempted to start UDP server, but it is already running.");
            return;
        }

        DatagramSocket serverSocket = new DatagramSocket(port);

        executor.submit(() -> {
            byte[] receiveData = new byte[RECEIVE_BUFFER_SIZE];

            while (true) {
                try {
                    handleRequest(serverSocket, receiveData);
                } catch (Exception e) {
                    logger.log(Level.SEVERE, "Failed to handle UDP request.", e);
                }
            }
        });

        started = true;
        logger.info("Started UDP server at port " + port);
    }

    private void handleRequest(final DatagramSocket serverSocket, final byte[] receiveData) throws IOException {
        DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
        serverSocket.receive(receivePacket);
        requestHandler.handle(serverSocket, receivePacket);
    }
}

File: udp-server/src/main/kotlin/io/sparkled/udpserver/impl/UdpServerImpl.kt
package io.sparkled.udpserver.impl

import io.sparkled.common.logging.getLogger
import io.sparkled.common.threading.NamedVirtualThreadFactory
import io.sparkled.udpserver.RequestHandler
import io.sparkled.udpserver.UdpServer
import jakarta.inject.Singleton
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.nio.charset.StandardCharsets
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

// TODO consider using NIO sockets.
@Singleton
class UdpServerImpl(
    private val requestHandler: RequestHandler,
) : UdpServer {

    private val executor: ExecutorService = Executors.newSingleThreadExecutor(
        NamedVirtualThreadFactory(javaClass.simpleName),
    )

    private val responseExecutor: ExecutorService = Executors.newFixedThreadPool(
        REQUEST_HANDLER_THREADS,
        NamedVirtualThreadFactory("${javaClass.simpleName}RequestHandler"),
    )

    private var started: Boolean = false

    override fun start(socket: DatagramSocket) {
        if (started) {
            logger.warn("Attempted to start UDP server, but it is already running.")
            return
        }

        started = true
        executor.submit { listen(socket) }

        logger.info("Started UDP server.", "port" to socket.localPort)
    }

    override fun stop() {
        started = false
    }

    private fun listen(serverSocket: DatagramSocket) {
        val receiveData = ByteArray(RECEIVE_BUFFER_SIZE)

        while (started) {
            try {
                handleRequest(serverSocket, receiveData)
            } catch (e: Exception) {
                logger.error("Failed to handle UDP request.", e)
            }
        }
    }

    private fun handleRequest(serverSocket: DatagramSocket, receiveData: ByteArray) {
        val receivePacket = DatagramPacket(receiveData, receiveData.size)
        serverSocket.receive(receivePacket)

        val message = String(receivePacket.data, 0, receivePacket.length, StandardCharsets.UTF_8)
        responseExecutor.submit { requestHandler.handle(serverSocket, receivePacket, message) }
    }

    companion object {
        private val logger = getLogger<UdpServerImpl>()
        private const val RECEIVE_BUFFER_SIZE = 32
        private const val REQUEST_HANDLER_THREADS = 4
    }
}
